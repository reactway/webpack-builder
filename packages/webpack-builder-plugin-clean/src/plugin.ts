import CleanWebpackPlugin from "clean-webpack-plugin";
import { Plugin } from "@reactway/webpack-builder";
import { Options } from "./plugin-options";

interface CleanPluginOptions {
    options?: Partial<Options>;
}

export const CleanPlugin: Plugin<CleanPluginOptions> = (config, projectDirectory) => webpack => {
    if (webpack.plugins == null) {
        webpack.plugins = [];
    }

    let cleanPluginConfig: Partial<Options> = {};
    if (config != null && config.options != null) {
        cleanPluginConfig = config.options;
    }

    webpack.plugins.push(new CleanWebpackPlugin({ ...cleanPluginConfig }));
    return webpack;
};
