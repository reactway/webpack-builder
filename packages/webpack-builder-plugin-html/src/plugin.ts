import HtmlWebpackPlugin from "html-webpack-plugin";
import HtmlWebpackTemplate from "html-webpack-template";
import { Plugin } from "@reactway/webpack-builder";

interface HtmlPluginOptions {
    options?: HtmlWebpackPlugin.Options;
}

export const HtmlPlugin: Plugin<HtmlPluginOptions> = (config, projectDirectory) => webpack => {
    if (webpack.plugins == null) {
        webpack.plugins = [];
    }

    let htmlPluginOptions: HtmlWebpackPlugin.Options | undefined = {
        inject: false,
        template: HtmlWebpackTemplate,
        baseHref: "/",
        appMountIds: ["root"],
        meta: [
            {
                name: "viewport",
                content: "width=device-width, initial-scale=1"
            }
        ]
    };

    if (config != null) {
        htmlPluginOptions = config.options;
    }

    // ts-jest: Throws error when it cannot resolve project root, because we are using Rush/pnpm.
    // tslint:disable-next-line:no-any
    webpack.plugins.push(new HtmlWebpackPlugin(htmlPluginOptions) as any);
    return webpack;
};
