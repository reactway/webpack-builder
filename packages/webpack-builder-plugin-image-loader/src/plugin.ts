import { Plugin as WebpackPlugin, Compiler } from "webpack";
import debug from "debug";
import chalk from "chalk";
import { ImageSizeData } from "./contracts";

const log = debug("rw-image-plugin");

log.enabled = true;

export class ReactwayImagePlugin implements WebpackPlugin {
    public static loader: string;
    public static imagesSizeArray: ImageSizeData[] = [];

    public apply(compiler: Compiler): void {
        // Chalk.
        const green = chalk.green;
        const white = chalk.white;
        const yellow = chalk.yellow;
        const blueBright = chalk.blueBright;

        compiler.hooks.done.tap("reactway-image-plugin", () => {
            const imagesSizeData = ReactwayImagePlugin.imagesSizeArray;
            if (imagesSizeData.length === 0) {
                return;
            }

            log("");
            log(green("RWIP | Images stats:".toLocaleUpperCase()));
            log(green("-".repeat(50)));

            imagesSizeData.forEach(image => {
                log("=".repeat(10));
                log(`| Raw image request: ${yellow(`${image.rawRequest}`)}`);
                log(`| Image file path: ${yellow(`${image.filePath}`)}`);
                log(`| Requested source file path: ${yellow(`${image.requestedFilePath}`)}`);
                if (image.isBase64 === true) {
                    log(blueBright("|Image converted to Base64."));
                    return;
                }

                if (image.reducedSize == null || image.originalSize.length === image.reducedSize.length) {
                    log(blueBright("| Reduced image size haven't changed compare to original image size."));
                    return;
                }

                const originalSize = white(`${image.originalSize.length}`);
                const reducedSize = white(`${image.reducedSize.length}`);

                // Image size compared to original in percentages.
                const reducedImagePercentage: number = 100 - (image.reducedSize.length * 100) / image.originalSize.length;
                const displayPercentage =
                    reducedImagePercentage > 0
                        ? chalk.greenBright(`${reducedImagePercentage.toFixed(2)}%`)
                        : chalk.redBright(`${reducedImagePercentage.toFixed(2)}%`);

                log(
                    blueBright(
                        // tslint:disable-next-line:max-line-length
                        `| Compared original image size(${originalSize}) with optimized/resized size(${reducedSize}) reduced by ${displayPercentage}`
                    )
                );
            });
            log("=".repeat(10));
            log(green("-".repeat(50)));
            log("");
        });
    }
}

ReactwayImagePlugin.loader = require.resolve("./image-loader");
