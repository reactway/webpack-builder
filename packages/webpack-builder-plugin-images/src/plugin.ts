import { Plugin } from "@reactway/webpack-builder";

const IMAGES_OUTPUT_LOCATION: string = "./assets/images";
const PUBLIC_PATH: string = "/";
// tslint:disable-next-line no-any
type OptionsDictionary = { [key: string]: any };

interface ImagesPluginOptions {
    imagesOutputLocation?: string;
    publicPath?: string;
    urlLoaderOptions?: OptionsDictionary;
}

export const ImagesPlugin: Plugin<ImagesPluginOptions> = (config, projectDirectory) => webpack => {
    if (webpack.module == null) {
        webpack.module = {
            rules: []
        };
    }

    const imagesOutputLocation: string =
        config != null && config.imagesOutputLocation != null ? config.imagesOutputLocation : IMAGES_OUTPUT_LOCATION;

    const publicPath: string = config != null && config.publicPath != null ? config.publicPath : PUBLIC_PATH;

    let urlLoaderOptions: OptionsDictionary = {};
    if (config != null && config.urlLoaderOptions != null) {
        urlLoaderOptions = config.urlLoaderOptions;
    }

    webpack.module.rules.push({
        test: /\.(png|jpg|gif|svg)$/,
        options: {
            name: `${imagesOutputLocation}/[name].[ext]`,
            publicPath: publicPath,
            limit: 10000,
            ...urlLoaderOptions
        },
        loader: "url-loader"
    });
    return webpack;
};
