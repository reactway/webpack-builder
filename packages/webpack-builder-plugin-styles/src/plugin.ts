import Webpack from "webpack";
import * as upath from "upath";
import * as fs from "fs-extra";
import queryString from "query-string";
import { Plugin } from "@reactway/webpack-builder";
import MiniCssExtractPlugin, { PluginOptions as MiniCssExtractPluginOptions } from "mini-css-extract-plugin";
import OptimizeCSSAssetsPlugin, { Options as OptimizeCSSOptions } from "optimize-css-assets-webpack-plugin";

// Extensions.
const CSS_EXTENSION: string = ".css";
const SCSS_EXTENSION: string = ".scss";

// Postcss.config.js
const POSTCSS_CONFIG_NAME: string = "postcss.config.js";

// Fonts location.
const FONTS_OUTPUT_LOCATION: string = "./assets/fonts";
// Public path
const PUBLIC_PATH: string = "./";

type Omit<TType, TKey extends keyof TType> = Pick<TType, Exclude<keyof TType, TKey>>;
type LoaderOptions = Omit<Webpack.RuleSetLoader, "loader">;

// Webpack.RuleSetQuery | undefined
type PostCssOptions = { [k: string]: any };
type PostCssPluginsFn = (...args: unknown[]) => unknown[];

interface StylesPluginOptions {
    fontsOutputLocation?: string;
    fontsPublicPath?: string;
    urlLoaderOptions?: LoaderOptions;
    styleLoaderOptions?: LoaderOptions;
    cssLoaderOptions?: LoaderOptions;
    postcssLoaderOptions?: LoaderOptions;
    sassLoaderOptions?: LoaderOptions;
    optimizeCssAssetsPlugin?: OptimizeCSSOptions;
    miniCssExtractPluginOptions?: MiniCssExtractPluginOptions;
}

export const StylesPlugin: Plugin<StylesPluginOptions> = (config, projectDirectory) => {
    checkPostCssConfig(projectDirectory);

    return webpack => {
        if (webpack.mode === "production") {
            if (webpack.plugins == null) {
                webpack.plugins = [];
            }

            let miniCssExtractPluginOptions: MiniCssExtractPluginOptions = {
                filename: "[name].[chunkhash].css",
                chunkFilename: "[name].[contenthash].css"
            };

            if (config != null && config.miniCssExtractPluginOptions != null) {
                miniCssExtractPluginOptions = config.miniCssExtractPluginOptions;
            }

            webpack.plugins.push(new MiniCssExtractPlugin(miniCssExtractPluginOptions) as Plugin);
        }

        if (webpack.module == null) {
            webpack.module = {
                rules: []
            };
        }

        const fontsOutputLocation: string =
            config != null && config.fontsOutputLocation != null ? config.fontsOutputLocation : FONTS_OUTPUT_LOCATION;

        const fontsPublicPath: string = config != null && config.fontsPublicPath != null ? config.fontsPublicPath : PUBLIC_PATH;

        const baseUrlLoaderOptions = {
            name: `${fontsOutputLocation}/[name].[ext]`,
            publicPath: fontsPublicPath,
            limit: 10000
        };

        const urlLoaderOptions: LoaderOptions = {
            options: {
                ...baseUrlLoaderOptions
            }
        };

        if (config != null && config.urlLoaderOptions != null) {
            const urlOptions = config.urlLoaderOptions.options;
            if (typeof urlOptions === "object") {
                urlLoaderOptions.options = {
                    ...baseUrlLoaderOptions,
                    ...urlOptions
                };
            }

            if (typeof urlOptions === "string") {
                urlLoaderOptions.options = urlOptions;
            }

            urlLoaderOptions.ident = config.urlLoaderOptions.ident;
            urlLoaderOptions.query = config.urlLoaderOptions.query;
        }

        let styleLoaderOptions: LoaderOptions = {};
        if (config != null && config.styleLoaderOptions != null) {
            styleLoaderOptions = config.styleLoaderOptions;
        }

        let cssLoaderOptions: LoaderOptions = {};
        if (config != null && config.cssLoaderOptions != null) {
            cssLoaderOptions = config.cssLoaderOptions;
        }

        let postcssLoaderOptions: LoaderOptions = {};
        if (config != null && config.postcssLoaderOptions != null) {
            postcssLoaderOptions = config.postcssLoaderOptions;
        }
        // Extract post-css options out loader options
        let postCssOptions: PostCssOptions = {};
        if (postcssLoaderOptions.options != null) {
            let options = postcssLoaderOptions.options;
            if (typeof options === "object") {
                postCssOptions = options;
            }

            if (typeof options === "string") {
                postCssOptions = queryString.parse(options);
            }
        }

        let postCssPlugins: PostCssPluginsFn | unknown[] | undefined = postCssOptions.plugins;
        // Default autoprefixer.
        const autoprefixerRequire = require("autoprefixer")();
        if (postCssPlugins == null) {
            postCssPlugins = [autoprefixerRequire];
        } else {
            const definedPostCssPlugins = postCssPlugins;
            if (Array.isArray(postCssPlugins)) {
                postCssPlugins = [...postCssPlugins, autoprefixerRequire];
            } else {
                const pluginsFn = definedPostCssPlugins as PostCssPluginsFn;
                postCssPlugins = (...args: unknown[]) => {
                    return [...pluginsFn(args), autoprefixerRequire];
                };
            }
        }

        postcssLoaderOptions.options = {
            ...postCssOptions,
            plugins: postCssPlugins
        };

        let sassLoaderOptions: LoaderOptions = {};
        if (config != null && config.sassLoaderOptions != null) {
            sassLoaderOptions = config.sassLoaderOptions;
        }

        const webpackRulesSet: Webpack.RuleSetUse = [
            {
                // Creates style nodes from JS strings.
                loader: "style-loader",
                ...styleLoaderOptions
            },
            {
                // "css-loader",
                loader: "css-loader",
                ...cssLoaderOptions
            },
            // // Translates CSS into CommonJS.
            // Autoprefixer
            {
                loader: "postcss-loader",
                ...postcssLoaderOptions
            },
            // Compiles Sass to CSS.
            {
                loader: "sass-loader",
                ...sassLoaderOptions
            }
        ];

        // If it is a `production` mode webpack then we need `MiniCssExtractPlugin` add after `styles` loader (it crashes if we put before).
        if (webpack.mode === "production") {
            webpackRulesSet.splice(1, 0, MiniCssExtractPlugin.loader);
        }

        webpack.module.rules.push(
            {
                test: /\.(sa|sc|c)ss$/,
                use: [...webpackRulesSet]
            },
            {
                test: /\.(woff|woff2|eot|ttf|otf)$/,
                use: [
                    {
                        loader: "url-loader",
                        ...urlLoaderOptions
                    }
                ]
            }
        );

        if (webpack.resolve == null) {
            webpack.resolve = {};
        }

        if (webpack.resolve.extensions == null) {
            webpack.resolve.extensions = [];
        }

        if (webpack.resolve.extensions.indexOf(CSS_EXTENSION) === -1) {
            webpack.resolve.extensions.push(CSS_EXTENSION);
        }

        if (webpack.resolve.extensions.indexOf(SCSS_EXTENSION) === -1) {
            webpack.resolve.extensions.push(SCSS_EXTENSION);
        }

        if (webpack.mode === "production") {
            if (webpack.optimization == null) {
                webpack.optimization = {};
            }

            if (webpack.optimization.minimizer == null) {
                webpack.optimization.minimizer = [];
            }

            webpack.optimization.minimizer.push(new OptimizeCSSAssetsPlugin(config == null ? undefined : config.optimizeCssAssetsPlugin));
        }

        return webpack;
    };
};

export function checkPostCssConfig(projectDirectory: string): void {
    const configLocation = upath.resolve(projectDirectory, POSTCSS_CONFIG_NAME);

    if (!fs.pathExistsSync(configLocation)) {
        console.warn(`${POSTCSS_CONFIG_NAME} was not found at project directory.`);
    }
}
