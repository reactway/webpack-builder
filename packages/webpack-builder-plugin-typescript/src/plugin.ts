import ForkTsCheckerWebpackPlugin from "fork-ts-checker-webpack-plugin";
import { TsconfigPathsPlugin } from "tsconfig-paths-webpack-plugin/lib";
import { Options as TsconfigPathsPluginOptions } from "tsconfig-paths-webpack-plugin/lib/options";
import { loadTsconfig, Tsconfig } from "tsconfig-paths/lib/tsconfig-loader";
import upath from "upath";
import * as fs from "fs-extra";
import { Plugin } from "@reactway/webpack-builder";
import { ForkTsCheckerWebpackPluginOptions } from "./plugin-options";

import "ts-loader";

// Extensions.
const TS_EXTENSION: string = ".ts";
const TSX_EXTENSION: string = ".tsx";
const JS_EXTENSION: string = ".js";
const JSX_EXTENSION: string = ".jsx";

// Tsconfig.
export const TS_CONFIG_NAME: string = "tsconfig.json";
const DEFAULT_TS_CONFIG_LOCATION: string = upath.resolve(__dirname, `../assets/${TS_CONFIG_NAME}`);

// TsLint.
export const TSLINT_CONFIG_NAME: string = "tslint.json";
const DEFAULT_TSLINT_CONFIG_LOCATION: string = upath.resolve(__dirname, `../assets/${TSLINT_CONFIG_NAME}`);

interface TypeScriptPluginOptions {
    forkTsCheckerOptions?: Partial<ForkTsCheckerWebpackPluginOptions>;
    tsconfigPathsPluginOptions?: Partial<TsconfigPathsPluginOptions>;
}

export const TypeScriptPlugin: Plugin<TypeScriptPluginOptions> = (config, projectDirectory) => {
    const fullTsconfigLocation = upath.resolve(projectDirectory, TS_CONFIG_NAME);
    let baseURLExist: boolean = false;

    try {
        checkTsConfig(projectDirectory);
    } catch (error) {
        console.error(`Failed while initiating "${TS_CONFIG_NAME}".`, error);
    }

    try {
        checkTslintConfig(projectDirectory);
    } catch (error) {
        console.error(`Failed while initiating "${TSLINT_CONFIG_NAME}".`, error);
    }

    const tsConfig: Tsconfig | undefined = loadTsconfig(fullTsconfigLocation);

    return webpack => {
        if (webpack.plugins == null) {
            webpack.plugins = [];
        }

        webpack.devtool = "inline-source-map";

        let forkTsConfig: Partial<ForkTsCheckerWebpackPluginOptions> = {};
        if (config != null && config.forkTsCheckerOptions != null) {
            forkTsConfig = config.forkTsCheckerOptions;
        }

        webpack.plugins.push(
            new ForkTsCheckerWebpackPlugin({
                checkSyntacticErrors: true,
                tslint: true,
                ...forkTsConfig
            })
        );

        if (webpack.module == null) {
            webpack.module = {
                rules: []
            };
        }

        webpack.module.rules.push({
            test: /\.tsx?$/,
            use: [
                {
                    loader: "babel-loader",
                    options: {
                        babelrc: true,
                        plugins: ["syntax-dynamic-import"]
                    }
                },
                {
                    loader: "ts-loader",
                    options: {
                        // disable type checker - we will use it in fork plugin
                        // IMPORTANT! use happyPackMode mode to speed-up compilation and reduce errors reported to webpack
                        happyPackMode: true,
                        transpileOnly: true
                    }
                }
            ],
            exclude: /node_modules/
        });

        webpack.module.rules.push({
            test: /\.jsx?$/,
            use: [
                {
                    loader: "babel-loader",
                    options: {
                        presets: ["@babel/preset-env"]
                    }
                }
            ],
            include: /node_modules/
        });

        if (webpack.resolve == null) {
            webpack.resolve = {};
        }

        if (webpack.resolve.plugins == null) {
            webpack.resolve.plugins = [];
        }

        if (
            tsConfig != null &&
            tsConfig.compilerOptions != null &&
            tsConfig.compilerOptions.baseUrl != null &&
            tsConfig.compilerOptions.baseUrl.trim().length !== 0
        ) {
            let configToSpread = {};
            if (config != null && config.tsconfigPathsPluginOptions != null) {
                configToSpread = config.tsconfigPathsPluginOptions;
            }

            const defaultOptions: Partial<TsconfigPathsPluginOptions> = {
                configFile: fullTsconfigLocation,
                ...configToSpread
            };

            webpack.resolve.plugins.push(new TsconfigPathsPlugin(defaultOptions));

            baseURLExist = true;
        }

        if (!baseURLExist && config != null && config.tsconfigPathsPluginOptions != null) {
            throw new Error(`Cannot add tsconfigPathsPluginOptions because baseUrl do not exist at ${TS_CONFIG_NAME}`);
        }

        if (webpack.resolve.extensions == null) {
            webpack.resolve.extensions = [];
        }

        if (webpack.resolve.extensions.indexOf(TS_EXTENSION) === -1) {
            webpack.resolve.extensions.push(TS_EXTENSION);
        }

        if (tsConfig != null && tsConfig.compilerOptions != null) {
            // HACK: To get other compiler options properties.
            // tslint:disable-next-line:no-any
            const tsConfigCompilerOptions = tsConfig.compilerOptions as any;
            if (tsConfigCompilerOptions.jsx != null) {
                if (webpack.resolve.extensions.indexOf(TSX_EXTENSION) === -1) {
                    webpack.resolve.extensions.push(TSX_EXTENSION);
                }
                if (webpack.resolve.extensions.indexOf(JSX_EXTENSION) === -1) {
                    webpack.resolve.extensions.push(JSX_EXTENSION);
                }
            }
        }

        if (webpack.resolve.extensions.indexOf(JS_EXTENSION) === -1) {
            webpack.resolve.extensions.push(JS_EXTENSION);
        }

        return webpack;
    };
};

export function checkTsConfig(projectDirectory: string): void {
    const configLocation = upath.resolve(projectDirectory, TS_CONFIG_NAME);

    if (!fs.pathExistsSync(configLocation)) {
        console.info(`File "${TS_CONFIG_NAME}" not found at ${configLocation}. Creating...`);
        fs.copySync(DEFAULT_TS_CONFIG_LOCATION, configLocation);
        console.info("Created.");
    }
}

export function checkTslintConfig(projectDirectory: string): void {
    const configLocation = upath.resolve(projectDirectory, TSLINT_CONFIG_NAME);

    if (!fs.pathExistsSync(configLocation)) {
        console.info(`File "${TSLINT_CONFIG_NAME}" not found at ${configLocation}. Creating...`);
        fs.copySync(DEFAULT_TSLINT_CONFIG_LOCATION, configLocation);
        console.info("Created.");
    }
}
