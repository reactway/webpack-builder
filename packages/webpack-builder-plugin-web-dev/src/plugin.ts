import * as path from "path";
import WebpackDevServer from "webpack-dev-server";

import { Plugin, Configuration } from "@reactway/webpack-builder";

const HOST: string = "0.0.0.0";
const DEFAULT_PORT: number = 3000;
const DEFAULT_OUTPUT_LOCATION: string = "./dist";

interface WebDevServerOptions {
    devServer?: WebpackDevServer.Configuration;
}

export const WebDevPlugin: Plugin<WebDevServerOptions> = (config, projectDirectory) => _webpack => {
    const webpackWithDevServer: Configuration & { devServer?: WebpackDevServer.Configuration } = _webpack;
    if (webpackWithDevServer.devServer == null) {
        webpackWithDevServer.devServer = {};
    }

    let webDevServer: WebpackDevServer.Configuration | undefined = {
        contentBase: path.resolve(projectDirectory, DEFAULT_OUTPUT_LOCATION),
        compress: true,
        host: HOST,
        quiet: false,
        port: DEFAULT_PORT,
        historyApiFallback: true
    };

    if (config != null) {
        webDevServer = config.devServer;
    }

    webpackWithDevServer.devServer = webDevServer;

    return webpackWithDevServer;
};
