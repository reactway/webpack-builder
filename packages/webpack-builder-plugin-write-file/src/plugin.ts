import WriteFileWebpackPlugin from "write-file-webpack-plugin";
import { Plugin } from "@reactway/webpack-builder";

export interface WriteFilePluginOptions {
    // tslint:disable-next-line no-any
    [key: string]: any;
}

export const WriteFilePlugin: Plugin<WriteFilePluginOptions> = (
    config: WriteFilePluginOptions | undefined,
    projectDirectory
) => webpack => {
    if (webpack.plugins == null) {
        webpack.plugins = [];
    }
    webpack.plugins.push(new WriteFileWebpackPlugin(config));
    return webpack;
};
