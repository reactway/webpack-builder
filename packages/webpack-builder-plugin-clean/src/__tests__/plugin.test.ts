import { Builder, Configuration } from "@reactway/webpack-builder";
import * as path from "path";
import { CleanPlugin } from "../plugin";

let SAMPLE_CONFIGURATION: Configuration = {};
const TEST_PROJECT_LOCATION: string = path.resolve(__dirname, "./test-project");

beforeEach(() => {
    SAMPLE_CONFIGURATION = {
        entry: "./src/index.ts",
        mode: "development",
        output: {
            path: path.resolve(TEST_PROJECT_LOCATION, "dist"),
            filename: "[name].bundle.js",
            chunkFilename: "[name].bundle.js",
            publicPath: "./"
        }
    };
});

it("Adding clean plugin to configuration", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION).use(CleanPlugin).toConfig();
    expect(configuration).toMatchSnapshot();
});

it("Adding clean plugin with options to configuration", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION).use(CleanPlugin, { options: { dry: true } }).toConfig();
    expect(configuration).toMatchSnapshot();
});
