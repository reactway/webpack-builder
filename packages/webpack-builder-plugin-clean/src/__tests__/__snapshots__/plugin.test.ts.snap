// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adding clean plugin to configuration 1`] = `
Object {
  "entry": "./src/index.ts",
  "mode": "development",
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": true,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
  ],
}
`;

exports[`Adding clean plugin with options to configuration 1`] = `
Object {
  "entry": "./src/index.ts",
  "mode": "development",
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": true,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": true,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": true,
    },
  ],
}
`;
